### PR TITLE
Add contract logs pagination

### LINKED ISSUE #3581: Add pagination to contract logs REST API
### Problem

The `/api/v1/contracts/{address}/results/logs` currently does not support pagination since it would've required a multi-column key.

### Solution

`/api/v1/contracts/{address}/results/logs` and `/api/v1/contracts/results/logs` API will be enhanced to support pagination. The existing `index` and `transaction` parameters will be used as the multi-column key and be set for subsequent requests in the `next` link. The `index` parameter cannot be present without the `timestamp` parameter. The index only applies within a given timestamp, not across timestamps. 

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by ar-conmit**
**Description**:
This PR adds pagination to contract logs
- Affected endpoints:
    - `contracts/{id}/results/logs`
    - `contracts/results/logs` 

**Related issue(s)**:

Fixes #3581 

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by Nana-EC — COMMENTED** [review]
1st re-review pass to address test issue highlighted

**Comment by Nana-EC** [comment]
> > There is an edge case that makes an endless loop:
> > if we have a dataset of `timestamp=A` with `index=0,1,2,3,4,5` and query `logs?index=gt:2&timestamp=gte:A&index=lt:3&timestamp:lte:A&limit=3`
> > Result: order `ASC` -> `timestamp=A` & `Index=0,1,2` order `DESC` -> `timestmap=A` & `Index=5,4,3`
> > The generated `next` link for both results would be the same(as the query) `index=gt:2&timestamp=gte:A&index=lt:3&timestamp:lte:A&limit=3`
> > the union query looks like:
> > `lower` - index > 2 & timestamp = A `inner` - timestamp > A & timestamp < A `upper` - index < 3 & timestamp = A
> > so, the `lower` part matches `index=3,4,5` and the `upper` part matches `index=2,1,0`. It always finds the same match.
> 
> ~I don't think it can cause endless loop.~
> 
> ~1. when ordering `ASC`, the next link will return `timestamp=A, index=3,4,5`, then the next link will return empty response~ ~2. when ordering `DESC`, the next link will return `timestamp=A, index=2,1,0`, then the next link will return empty response~
> 
> I see the problem now, then we should convert the condition of `timestamp=gte:A` and `timestamp=lte:A` to `timestamp=A`

I think we should be rejecting such a query for now and suggesting the user simplify it
`logs?index=gt:2&timestamp=gte:A&index=lt:3&timestamp:lte:A&limit=3` should be `logs?index=gt:2&timestamp=gte:A&index=lt:3&limit=3`

File: hedera-mirror-rest/controllers/contractController.js [hunk 1]
```diff
@@ -552,14 +553,73 @@ class ContractController extends BaseController {
...
+      (!timestampBound.hasUpper() || timestampBound.upper.operator === utils.opsMap.lt)
+    ) {
+      // invalid ops: index </<= & timestamp >/>=/<
+      throw new InvalidArgumentError(`Timestamp must have lte or eq operator`);
+    }
+
+    if (indexBound.hasEqual() && !timestampBound.hasEqual()) {
+      throw new InvalidArgumentError(`Timestamp must have eq operator`);
+    }
+  };
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/controllers/contractController.js:591]: works for me

[the remark this answers is not in the document]
Comment (ar-conmit) [reply hedera-mirror-rest/controllers/contractController.js:591]: If there are more records to be fetched, the `next` link sometimes will include 2x timestamps with the same value and the above check will reject it which is not the best user experience.

I think the easiest way of preventing this endless loop is just to not sending a `next` link.
We will generate the `next` link and if its query params are identical to the requested ones, we should not return the next link

[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/controllers/contractController.js:591]: oh good catch. Then I'd suggest what I proposed at first, if we end up with `timestamp=lte:A&timestamp=gte:A`, update the `bound` object as if it was `timestamp=A`

File: hedera-mirror-rest/__tests__/specs/contracts-logs-24-pagination-index-timestamp-range.spec copy.json [hunk 2]
```diff
@@ -0,0 +1,252 @@
+{
```
Comment (xin-hedera) [inline hedera-mirror-rest/__tests__/specs/contracts-logs-24-pagination-index-timestamp-range.spec copy.json:1]: rename file remove the `copy` in the filename

**Review by xin-hedera — CHANGES_REQUESTED** [review]
looks good. just the open item regarding how to handle `timestamp=gte:A&timestamp=lte:A`

**Review by Nana-EC — APPROVED** [review]
Looks good.
I'm unblocking the PR but would like a follow up ticket opened up to investigate the desired response over all since this can happen in other places.

I think we should reject and leave it on the user for now where there's no applicable subset.
We can internally resolve where the values match as done here and apply the logic to more endpoints.

File: hedera-mirror-rest/__tests__/controllers/contractController.test.js [hunk 3]
```diff
@@ -1538,3 +1538,71 @@ describe('extractContractLogsMultiUnionQuery - negative', () => {
...
+    },
+    {
+      name: 'timestamp = 34',
+      input: {
+        req: {query: {...inputQuery, timestamp: timestamp34}},
+      },
+      expected: {query: {...expectedQuery, timestamp: timestamp34}},
+    },
+    {
+      name: 'timestamp => 34 & timestamp < 34',
```
[the remark this answers is not in the document]
Comment (ar-conmit) [reply hedera-mirror-rest/__tests__/controllers/contractController.test.js:1566]: Agree, I am just testing the new function of overwriting the timestamp query params, not the whole flow.

**Review by xin-hedera — APPROVED** [review]
LGTM

**Review by matheus-dallrosa — APPROVED** [review]

**Review by steven-sheehy — APPROVED** [review]
LGTM. Just some leftovers in the openapi that you can address via your fee schedule PR or another followup.

**Pull request merged**